src/utils.jl

```diff
@@ -0,0 +1,59 @@
+struct SolverFailureException{A,B} <: Exception
+    solver::A
+    stats::B
+end
+
+function Base.show(io::IO, sfe::SolverFailureException)
+    return println(
+        io,
+        "SolverFailureException: \n Solver: $(sfe.solver) \n Solver stats: $(string(sfe.stats))",
+    )
+end
+
+function check_solution(solver, stats)
+    if stats.solved
+        return nothing
+    else
+        throw(SolverFailureException(solver, stats))
+    end
+end
+
+"""
+    PushforwardMul!{P,N}
```

Comment:
Replaced `lazy_jacobian` with `pushforward/pullback_function` due to https://github.com/JuliaDiff/AbstractDifferentiation.jl/issues/83